File: julia/06-io-operations/io_operations.jl
using Printf

# --- Writing a file (do-block auto-closes it) ---
open("notes.txt", "w") do file
    println(file, "Julia I/O Demo")
    println(file, "Line 2 of 3")
    println(file, "Line 3 of 3")
end
println("File written.")

# --- Reading the whole file at once ---
content = read("notes.txt", String)
println("\nFull contents:")
print(content)

# --- Reading line by line with enumerate for line numbers ---
println("\nNumbered lines:")
for (i, line) in enumerate(eachline("notes.txt"))
    @printf("%d: %s\n", i, line)
end

# --- Appending in "a" mode does not overwrite ---
open("notes.txt", "a") do file
    println(file, "Appended line")
end

lines = readlines("notes.txt")
println("\nTotal lines now: $(length(lines))")

File: lua/06-io-operations/io_operations.lua
-- Open for writing ("w" creates the file or truncates an existing one)
local file = io.open("notes.txt", "w")
file:write("Lua I/O Demo\n")
file:write("Line 2 of 3\n")
file:write(string.format("Line %d of 3\n", 3))
file:close()
print("File written.")

-- Read the whole file at once with the "a" (all) format
local whole = io.open("notes.txt", "r")
local content = whole:read("a")
whole:close()
print("\nFull contents:")
io.write(content)

-- Iterate line by line; io.lines opens and closes the file for you
print("\nNumbered lines:")
local n = 0
for line in io.lines("notes.txt") do
    n = n + 1
    print(string.format("%d: %s", n, line))
end

-- Append mode ("a") adds to the end without overwriting
local append = io.open("notes.txt", "a")
append:write("Appended line\n")
append:close()

-- Count the lines now that we have appended one
local count = 0
for _ in io.lines("notes.txt") do
    count = count + 1
end
print("\nTotal lines now: " .. count)
